File: frontend/src/api/candidate.ts
/**
 * Candidate API — SS2
 * Backend mounted at /api/v1/candidates
 */
import { api } from './axios'

export interface CandidateEligibility {
  id: number
  user_id: number
  source: 'auto_certificate' | 'spoc_approved' | string
  eligible: boolean
  reason?: string
  decided_at?: string | null
  cohort_id?: number | null
}

export interface CandidateProfile {
  id: number
  user_id: number
  is_visible: boolean
  resume_url?: string
  bio?: string
  skills: string[]
  preferred_roles: string[]
  availability_date?: string | null
  linkedin_url?: string
  github_url?: string
  portfolio_url?: string
  created_at?: string
  updated_at?: string
  completeness: number
  eligibility?: CandidateEligibility | null
}

export interface CandidateProfileUpdate {
  resume_url?: string
  bio?: string
  skills?: string[]
  preferred_roles?: string[]
  availability_date?: string | null
  linkedin_url?: string
  github_url?: string
  portfolio_url?: string
  is_visible?: boolean
}

export interface CandidateSummary {
  user_id: number
  display_name: string
  bio?: string
  skills: string[]
  preferred_roles: string[]
  availability_date?: string | null
  linkedin_url?: string
  github_url?: string
  portfolio_url?: string
  eligibility_source?: string | null
}

export interface CandidateBrowseResponse {
  total: number
  page: number
  page_size: number
  items: CandidateSummary[]
}

export const candidateAPI = {
  async getMe(): Promise<CandidateProfile> {
    const res = await api.get<CandidateProfile>('/candidates/me')
    return res.data
  },

  async updateMe(payload: CandidateProfileUpdate): Promise<CandidateProfile> {
    const res = await api.put<CandidateProfile>('/candidates/me', payload)
    return res.data
  },

  async setVisibility(is_visible: boolean): Promise<CandidateProfile> {
    const res = await api.post<CandidateProfile>('/candidates/me/visibility', {
      is_visible,
    })
    return res.data
  },

  async browse(params: {
    skills?: string
    roles?: string
    availability_before?: string
    page?: number
    page_size?: number
  } = {}): Promise<CandidateBrowseResponse> {
    const res = await api.get<CandidateBrowseResponse>('/candidates/browse', { params })
    return res.data
  },

  async getDetail(userId: number) {
    const res = await api.get(`/candidates/${userId}`)
    return res.data
  },
}

export default candidateAPI

File: backend/app/schemas/candidate.py
"""
Candidate schemas — SS2. Pydantic v2.
"""
from __future__ import annotations

from datetime import date, datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field, HttpUrl


class CandidateProfileBase(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    is_visible: bool = False
    resume_url: Optional[str] = ""
    bio: Optional[str] = ""
    skills: List[str] = Field(default_factory=list)
    preferred_roles: List[str] = Field(default_factory=list)
    availability_date: Optional[date] = None
    linkedin_url: Optional[str] = ""
    github_url: Optional[str] = ""
    portfolio_url: Optional[str] = ""


class CandidateProfileUpdate(BaseModel):
    """Partial update — every field optional."""
    model_config = ConfigDict(from_attributes=True)

    resume_url: Optional[str] = None
    bio: Optional[str] = None
    skills: Optional[List[str]] = None
    preferred_roles: Optional[List[str]] = None
    availability_date: Optional[date] = None
    linkedin_url: Optional[str] = None
    github_url: Optional[str] = None
    portfolio_url: Optional[str] = None
    is_visible: Optional[bool] = None


class VisibilityToggle(BaseModel):
    is_visible: bool


class CandidateEligibilityOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    source: str
    eligible: bool
    reason: Optional[str] = ""
    decided_at: Optional[datetime] = None
    cohort_id: Optional[int] = None


class CandidateProfileOut(CandidateProfileBase):
    id: int
    user_id: int
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None
    completeness: int = 0
    eligibility: Optional[CandidateEligibilityOut] = None


class CandidateSummary(BaseModel):
    """Compact view returned in /browse."""
    model_config = ConfigDict(from_attributes=True)

    user_id: int
    display_name: str
    bio: Optional[str] = ""
    skills: List[str] = Field(default_factory=list)
    preferred_roles: List[str] = Field(default_factory=list)
    availability_date: Optional[date] = None
    linkedin_url: Optional[str] = ""
    github_url: Optional[str] = ""
    portfolio_url: Optional[str] = ""
    eligibility_source: Optional[str] = None


class CandidateBrowseResponse(BaseModel):
    total: int
    page: int
    page_size: int
    items: List[CandidateSummary]


class CandidateDetailResponse(CandidateSummary):
    resume_url: Optional[str] = ""
    # Email/phone are NOT exposed on the detail view — they are only
    # revealed after an accepted CompanyInterest (SS3).
